File: scripts/md-to-docx.py
#!/usr/bin/env python3
"""
Convertit la note méthodologique Markdown en document Word professionnel.

Sous-ensemble Markdown pris en charge : titres # à ####, paragraphes,
tableaux GFM, listes à puces et numérotées, blocs de code, citations,
règles horizontales, et le formatage en ligne gras / italique / code.

Usage : python3 scripts/md-to-docx.py entree.md sortie.docx "Titre" "Sous-titre"
"""
import re
import sys

from docx import Document
from docx.enum.section import WD_SECTION
from docx.enum.table import WD_TABLE_ALIGNMENT
from docx.enum.text import WD_ALIGN_PARAGRAPH, WD_BREAK
from docx.oxml import OxmlElement
from docx.oxml.ns import qn
from docx.shared import Inches, Pt, RGBColor

BRAND = RGBColor(0x0F, 0x3D, 0x63)      # bleu institutionnel
GREY = RGBColor(0x5C, 0x64, 0x70)
HEADER_FILL = "0F3D63"
BAND_FILL = "F2F5F8"


# --------------------------------------------------------------------------
# Styles
# --------------------------------------------------------------------------
def setup_styles(doc: Document) -> None:
    normal = doc.styles["Normal"]
    normal.font.name = "Calibri"
    normal.font.size = Pt(10)
    normal.paragraph_format.space_after = Pt(6)
    normal.paragraph_format.line_spacing = 1.15

    specs = [
        ("Heading 1", 18, True, BRAND, 20, 10),
        ("Heading 2", 14, True, BRAND, 16, 8),
        ("Heading 3", 11.5, True, RGBColor(0x1D, 0x1D, 0x1D), 12, 6),
        ("Heading 4", 10.5, True, GREY, 10, 4),
    ]
    for name, size, bold, color, before, after in specs:
        st = doc.styles[name]
        st.font.name = "Calibri"
        st.font.size = Pt(size)
        st.font.bold = bold
        st.font.color.rgb = color
        st.paragraph_format.space_before = Pt(before)
        st.paragraph_format.space_after = Pt(after)
        st.paragraph_format.keep_with_next = True

    code = doc.styles.add_style("CodeBlock", 1)
    code.font.name = "Consolas"
    code.font.size = Pt(8.5)
    code.paragraph_format.space_before = Pt(6)
    code.paragraph_format.space_after = Pt(6)
    code.paragraph_format.left_indent = Inches(0.25)


def shade(cell, fill: str) -> None:
    el = OxmlElement("w:shd")
    el.set(qn("w:val"), "clear")
    el.set(qn("w:fill"), fill)
    cell._tc.get_or_add_tcPr().append(el)


def set_repeat_header(row) -> None:
    tr_pr = row._tr.get_or_add_trPr()
    el = OxmlElement("w:tblHeader")
    el.set(qn("w:val"), "true")
    tr_pr.append(el)


# --------------------------------------------------------------------------
# Formatage en ligne
# --------------------------------------------------------------------------
INLINE = re.compile(r"(\*\*.+?\*\*|`[^`]+`|\*[^*]+\*)")


def add_inline(paragraph, text: str, base_bold=False, base_italic=False):
    # Liens Markdown : conserver le libellé, écarter l'URL.
    text = re.sub(r"\[([^\]]+)\]\([^)]+\)", r"\1", text)
    text = text.replace("<br>", " ").replace("\\|", "|")
    for part in INLINE.split(text):
        if not part:
            continue
        if part.startswith("**") and part.endswith("**") and len(part) > 4:
            run = paragraph.add_run(part[2:-2])
            run.bold = True
        elif part.startswith("`") and part.endswith("`") and len(part) > 2:
            run = paragraph.add_run(part[1:-1])
            run.font.name = "Consolas"
            run.font.size = Pt(9)
            run.font.color.rgb = RGBColor(0x8A, 0x2B, 0x2B)
        elif part.startswith("*") and part.endswith("*") and len(part) > 2:
            run = paragraph.add_run(part[1:-1])
            run.italic = True
        else:
            run = paragraph.add_run(part)
        if base_bold:
            run.bold = True
        if base_italic:
            run.italic = True
    return paragraph


# --------------------------------------------------------------------------
# Page de garde et table des matières
# --------------------------------------------------------------------------
def cover_page(doc: Document, title: str, subtitle: str) -> None:
    for _ in range(6):
        doc.add_paragraph()

    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    r = p.add_run(title)
    r.font.size = Pt(26)
    r.font.bold = True
    r.font.color.rgb = BRAND

    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    r = p.add_run(subtitle)
    r.font.size = Pt(14)
    r.font.color.rgb = GREY

    doc.add_paragraph()
    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    pf = p.paragraph_format
    pf.space_before = Pt(12)
    for line in [
        "Version 2.0 — 19 août 2026",
        "Direction des Risques — Modèles de crédit entreprises",
        "Usage interne",
    ]:
        r = p.add_run(line + "\n")
        r.font.size = Pt(10.5)
        r.font.color.rgb = GREY

    for _ in range(8):
        doc.add_paragraph()

    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    r = p.add_run(
        "Document de travail. Les pondérations, seuils et barèmes présentés ne constituent "
        "ni des seuils réglementaires, ni des probabilités de défaut calibrées, ni une "
        "validation de modèle."
    )
    r.font.size = Pt(8.5)
    r.italic = True
    r.font.color.rgb = GREY

    doc.add_paragraph().add_run().add_break(WD_BREAK.PAGE)


def toc_field(doc: Document) -> None:
    p = doc.add_paragraph()
    r = p.add_run("Table des matières")
    r.font.size = Pt(16)
    r.bold = True
    r.font.color.rgb = BRAND
    doc.add_paragraph()

    p = doc.add_paragraph()
    run = p.add_run()
    fld = OxmlElement("w:fldChar")
    fld.set(qn("w:fldCharType"), "begin")
    instr = OxmlElement("w:instrText")
    instr.set(qn("xml:space"), "preserve")
    instr.text = r'TOC \o "1-3" \h \z \u'
    sep = OxmlElement("w:fldChar")
    sep.set(qn("w:fldCharType"), "separate")
    placeholder = OxmlElement("w:t")
    placeholder.text = "Ouvrir dans Word puis actualiser le champ (F9) pour générer la table."
    end = OxmlElement("w:fldChar")
    end.set(qn("w:fldCharType"), "end")
    for el in (fld, instr, sep, placeholder, end):
        run._r.append(el)

    doc.add_paragraph().add_run().add_break(WD_BREAK.PAGE)


def page_numbers(doc: Document) -> None:
    footer = doc.sections[0].footer
    p = footer.paragraphs[0]
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = p.add_run()
    for tag, attr, text in (
        ("w:fldChar", "begin", None),
        ("w:instrText", None, "PAGE"),
        ("w:fldChar", "end", None),
    ):
        el = OxmlElement(tag)
        if attr:
            el.set(qn("w:fldCharType"), attr)
        if text:
            el.set(qn("xml:space"), "preserve")
            el.text = text
        run._r.append(el)
    run.font.size = Pt(8.5)
    run.font.color.rgb = GREY


# --------------------------------------------------------------------------
# Tableaux
# --------------------------------------------------------------------------
def split_row(line: str):
    line = line.strip()
    if line.startswith("|"):
        line = line[1:]
    if line.endswith("|"):
        line = line[:-1]
    # Sépare sur | non échappé.
    return [c.strip() for c in re.split(r"(?<!\\)\|", line)]


def add_table(doc: Document, rows, usable_width_emu: int) -> None:
    header = rows[0]
    body = rows[1:]
    ncols = len(header)
    table = doc.add_table(rows=1, cols=ncols)
    table.style = "Table Grid"
    table.alignment = WD_TABLE_ALIGNMENT.CENTER
    table.autofit = False

    # Largeurs proportionnelles au contenu, bornées pour rester lisibles.
    weights = []
    for i in range(ncols):
        cells = [header[i]] + [r[i] if i < len(r) else "" for r in body]
        longest = max((len(c) for c in cells), default = 1)
        weights.append(min(max(longest, 6), 60))
    total = sum(weights)
    widths = [int(usable_width_emu * w / total) for w in weights]
    table.columns_width = widths

    hdr = table.rows[0]
    set_repeat_header(hdr)
    for i, text in enumerate(header):
        cell = hdr.cells[i]
        cell.width = widths[i]
        cell.text = ""
        para = cell.paragraphs[0]
        para.paragraph_format.space_after = Pt(2)
        para.paragraph_format.space_before = Pt(2)
        add_inline(para, text, base_bold=True)
        for run in para.runs:
            run.font.size = Pt(8.5)
            run.font.color.rgb = RGBColor(0xFF, 0xFF, 0xFF)
        shade(cell, HEADER_FILL)

    for idx, row in enumerate(body):
        cells = table.add_row().cells
        for i in range(ncols):
            cell = cells[i]
            cell.width = widths[i]
            cell.text = ""
            para = cell.paragraphs[0]
            para.paragraph_format.space_after = Pt(2)
            para.paragraph_format.space_before = Pt(2)
            add_inline(para, row[i] if i < len(row) else "")
            for run in para.runs:
                run.font.size = Pt(8.5)
            if idx % 2 == 1:
                shade(cell, BAND_FILL)

    doc.add_paragraph().paragraph_format.space_after = Pt(4)


# --------------------------------------------------------------------------
# Conversion principale
# --------------------------------------------------------------------------
def convert(md_path: str, out_path: str, title: str, subtitle: str) -> None:
    lines = open(md_path, encoding="utf-8").read().split("\n")

    doc = Document()
    section = doc.sections[0]
    section.page_width = Inches(8.27)   # A4
    section.page_height = Inches(11.69)
    section.left_margin = Inches(0.85)
    section.right_margin = Inches(0.85)
    section.top_margin = Inches(0.8)
    section.bottom_margin = Inches(0.8)
    usable = section.page_width - section.left_margin - section.right_margin

    setup_styles(doc)
    cover_page(doc, title, subtitle)
    toc_field(doc)
    page_numbers(doc)

    i = 0
    n = len(lines)
    in_code = False
    code_buffer = []
    skip_toc_block = False

    while i < n:
        line = lines[i]
        stripped = line.strip()

        # Blocs de code
        if stripped.startswith("```"):
            if in_code:
                para = doc.add_paragraph(style="CodeBlock")
                para.add_run("\n".join(code_buffer))
                shade_para(para)
                code_buffer = []
                in_code = False
            else:
                in_code = True
            i += 1
            continue
        if in_code:
            code_buffer.append(line)
            i += 1
            continue

        # Le sommaire rédigé à la main est remplacé par le champ TOC de Word.
        if stripped == "# Sommaire":
            skip_toc_block = True
            i += 1
            continue
        if skip_toc_block:
            if stripped.startswith("---"):
                skip_toc_block = False
            i += 1
            continue

        if not stripped:
            i += 1
            continue

        # Règle horizontale
        if re.fullmatch(r"-{3,}", stripped):
            para = doc.add_paragraph()
            pf = para.paragraph_format
            pf.space_before = Pt(4)
            pf.space_after = Pt(8)
            pbdr = OxmlElement("w:pBdr")
            bottom = OxmlElement("w:bottom")
            bottom.set(qn("w:val"), "single")
            bottom.set(qn("w:sz"), "6")
            bottom.set(qn("w:color"), "D6DBE1")
            pbdr.append(bottom)
            para._p.get_or_add_pPr().append(pbdr)
            i += 1
            continue

        # Titres
        m = re.match(r"^(#{1,4})\s+(.*)$", stripped)
        if m:
            level = len(m.group(1))
            text = m.group(2)
            if level == 1:
                doc.add_paragraph().add_run().add_break(WD_BREAK.PAGE)
            para = doc.add_paragraph(style=f"Heading {level}")
            add_inline(para, text)
            i += 1
            continue

        # Tableaux
        if stripped.startswith("|") and i + 1 < n and re.match(
            r"^\s*\|?[\s:|-]+\|[\s:|-]*$", lines[i + 1]
        ):
            rows = [split_row(stripped)]
            i += 2
            while i < n and lines[i].strip().startswith("|"):
                rows.append(split_row(lines[i]))
                i += 1
            add_table(doc, rows, usable)
            continue

        # Citations
        if stripped.startswith(">"):
            buffer = []
            while i < n and lines[i].strip().startswith(">"):
                buffer.append(lines[i].strip().lstrip(">").strip())
                i += 1
            para = doc.add_paragraph()
            pf = para.paragraph_format
            pf.left_indent = Inches(0.25)
            pf.space_before = Pt(6)
            pf.space_after = Pt(8)
            pbdr = OxmlElement("w:pBdr")
            left = OxmlElement("w:left")
            left.set(qn("w:val"), "single")
            left.set(qn("w:sz"), "18")
            left.set(qn("w:space"), "8")
            left.set(qn("w:color"), "0F3D63")
            pbdr.append(left)
            para._p.get_or_add_pPr().append(pbdr)
            add_inline(para, " ".join(x for x in buffer if x))
            for run in para.runs:
                run.font.size = Pt(9.5)
            i += 1 if i < n else 0
            continue

        # Listes
        m = re.match(r"^(\s*)[-*]\s+(.*)$", line)
        if m:
            para = doc.add_paragraph(style="List Bullet")
            para.paragraph_format.space_after = Pt(2)
            if len(m.group(1)) >= 2:
                para.paragraph_format.left_indent = Inches(0.6)
            add_inline(para, m.group(2))
            i += 1
            continue

        m = re.match(r"^(\s*)\d+\.\s+(.*)$", line)
        if m:
            para = doc.add_paragraph(style="List Number")
            para.paragraph_format.space_after = Pt(2)
            add_inline(para, m.group(2))
            i += 1
            continue

        # Paragraphe : agrège les lignes consécutives.
        buffer = [stripped]
        i += 1
        while i < n:
            nxt = lines[i].strip()
            if (
                not nxt
                or nxt.startswith(("#", "|", ">", "```", "- ", "* "))
                or re.match(r"^\d+\.\s", nxt)
                or re.fullmatch(r"-{3,}", nxt)
            ):
                break
            buffer.append(nxt)
            i += 1
        add_inline(doc.add_paragraph(), " ".join(buffer))

    doc.save(out_path)
    print(f"Document généré : {out_path}")


def shade_para(para) -> None:
    el = OxmlElement("w:shd")
    el.set(qn("w:val"), "clear")
    el.set(qn("w:fill"), "F4F6F8")
    para._p.get_or_add_pPr().append(el)


if __name__ == "__main__":
    if len(sys.argv) < 5:
        print(__doc__)
        sys.exit(1)
    convert(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4])

File: scripts/md-to-docx.mjs
/**
 * Convertit un document Markdown du dépôt en fichier Word.
 *
 * Écrit pour les notes et argumentaires de `docs/` : titres, tableaux,
 * citations, blocs de code et emphase. Aucune dépendance à un service externe.
 *
 * La page de garde est construite depuis l'en-tête du document source — titre,
 * champs « **Nom :** valeur » et bloc de citation d'avertissement — de sorte
 * qu'une correction dans le Markdown se répercute dans le Word sans édition
 * manuelle. C'est la même règle que pour les grilles du modèle : une seule
 * source, jamais de recopie.
 *
 * Usage :
 *   node scripts/md-to-docx.mjs <source.md> <sortie.docx> [titre] [sous-titre]
 *                               [accroche] [version] [en-tête de page]
 */
import fs from "node:fs";
import path from "node:path";
import {
  Document,
  Packer,
  Paragraph,
  TextRun,
  HeadingLevel,
  Table,
  TableRow,
  TableCell,
  WidthType,
  ShadingType,
  BorderStyle,
  AlignmentType,
  PageBreak,
  Header,
  Footer,
  PageNumber,
  VerticalAlign,
  TabStopType,
  TabStopPosition,
} from "docx";

const [, , srcArg, outArg, ...coverArgs] = process.argv;
if (!srcArg || !outArg) {
  console.error(
    "Usage : node scripts/md-to-docx.mjs <source.md> <sortie.docx> [titre] [sous-titre] [accroche] [version] [en-tête]"
  );
  process.exit(1);
}

const SRC = path.resolve(srcArg);
const OUT = path.resolve(outArg);
const raw = fs.readFileSync(SRC, "utf8");
const lines = raw.split("\n");

const COVER = {
  title: coverArgs[0] ?? firstHeading() ?? path.basename(SRC, ".md"),
  subtitle: coverArgs[1] ?? "",
  tagline: coverArgs[2] ?? "",
  version: coverArgs[3] ?? "",
  header: coverArgs[4] ?? COVER_HEADER_FALLBACK(),
};

function COVER_HEADER_FALLBACK() {
  return coverArgs[0] ?? firstHeading() ?? "";
}

function firstHeading() {
  const h = lines.find((l) => /^# /.test(l));
  return h ? h.replace(/^#\s+/, "").trim() : null;
}

const BRAND = "1F3864";
const BRAND2 = "2E5395";
const LIGHT = "EDF2FA";
const FONT = "Calibri";
const CONTENT_WIDTH = 9600; // dxa

// ---------------------------------------------------------------------------
// Markdown en ligne
// ---------------------------------------------------------------------------

/**
 * Découpe le texte en fragments {text, bold, italics, code}.
 *
 * Renvoie des descripteurs simples, JAMAIS des TextRun : c'est l'appelant —
 * paragraphe ou cellule — qui applique sa taille et sa couleur de base. Un
 * TextRun déjà construit ne se relit pas, et tenter de le faire produit des
 * cellules vides.
 */
function parseInline(text) {
  const frags = [];
  let i = 0;
  let buf = "";
  const flush = () => {
    if (buf.length) frags.push({ text: buf });
    buf = "";
  };
  while (i < text.length) {
    if (text.startsWith("**", i)) {
      const end = text.indexOf("**", i + 2);
      if (end !== -1) {
        flush();
        frags.push({ text: text.slice(i + 2, end), bold: true });
        i = end + 2;
        continue;
      }
    }
    if (text[i] === "*" && text[i + 1] !== "*" && (i === 0 || text[i - 1] !== "*")) {
      const end = text.indexOf("*", i + 1);
      if (end !== -1 && end > i + 1) {
        flush();
        frags.push({ text: text.slice(i + 1, end), italics: true });
        i = end + 1;
        continue;
      }
    }
    if (text[i] === "`") {
      const end = text.indexOf("`", i + 1);
      if (end !== -1) {
        flush();
        frags.push({ text: text.slice(i + 1, end), code: true });
        i = end + 1;
        continue;
      }
    }
    buf += text[i];
    i += 1;
  }
  flush();
  return frags.length ? frags : [{ text: "" }];
}

function toRuns(fragments, base = {}) {
  return fragments.map((f) => {
    if (f.code) {
      return new TextRun({
        text: f.text,
        font: "Consolas",
        size: Math.max(16, (base.size || 21) - 2),
        shading: { type: ShadingType.CLEAR, fill: "F2F2F2" },
        color: base.color,
      });
    }
    return new TextRun({
      text: f.text,
      font: FONT,
      ...base,
      bold: f.bold || base.bold,
      italics: f.italics || base.italics,
    });
  });
}

// ---------------------------------------------------------------------------
// Tableaux
// ---------------------------------------------------------------------------

function parseAlign(sep) {
  const t = sep.trim();
  if (t.startsWith(":") && t.endsWith(":")) return AlignmentType.CENTER;
  if (t.endsWith(":")) return AlignmentType.RIGHT;
  return AlignmentType.LEFT;
}

function splitRow(line) {
  let s = line.trim();
  if (s.startsWith("|")) s = s.slice(1);
  if (s.endsWith("|")) s = s.slice(0, -1);
  return s.split("|").map((c) => c.trim());
}

function longestWord(s) {
  return s
    .replace(/\*\*/g, "")
    .replace(/`/g, "")
    .split(/\s+/)
    .reduce((m, w) => Math.max(m, w.length), 0);
}

/**
 * Largeurs de colonnes.
 *
 * Deux mesures par colonne : la longueur totale, qui donne la proportion, et le
 * plus long mot insécable, qui impose un plancher. Sans ce plancher, un jeton
 * comme « CREDIT_POLICY » se coupe en plein milieu — il ne contient aucune
 * espace où le rendu puisse revenir à la ligne.
 */
function buildTable(headerCells, aligns, bodyRows) {
  const n = headerCells.length;
  const maxLen = new Array(n).fill(0);
  const maxWord = new Array(n).fill(0);
  const measure = (cells) =>
    cells.forEach((c, idx) => {
      maxLen[idx] = Math.max(maxLen[idx] || 0, c.length);
      maxWord[idx] = Math.max(maxWord[idx] || 0, longestWord(c));
    });
  measure(headerCells);
  bodyRows.forEach(measure);

  const rawWidths = maxLen.map((l) => Math.max(700, Math.min(l * 55, 3000)));
  const total = rawWidths.reduce((a, b) => a + b, 0);
  const widths = rawWidths.map((w) => Math.round((w / total) * CONTENT_WIDTH));
  widths[widths.length - 1] += CONTENT_WIDTH - widths.reduce((a, b) => a + b, 0);

  const floors = maxWord.map((w) => Math.min(2700, Math.max(1000, w * 118 + 300)));
  let deficit = 0;
  for (let idx = 0; idx < widths.length; idx += 1) {
    if (widths[idx] < floors[idx]) {
      deficit += floors[idx] - widths[idx];
      widths[idx] = floors[idx];
    }
  }
  while (deficit > 0) {
    const donor = widths.reduce(
      (best, w, idx) => (w - floors[idx] > widths[best] - floors[best] ? idx : best),
      0
    );
    const room = widths[donor] - floors[donor];
    if (room <= 0) break;
    const take = Math.min(room, deficit);
    widths[donor] -= take;
    deficit -= take;
  }

  const cell = (text, { header = false, align = AlignmentType.LEFT, width }) =>
    new TableCell({
      width: { size: width, type: WidthType.DXA },
      shading: header ? { type: ShadingType.CLEAR, fill: BRAND } : undefined,
      verticalAlign: VerticalAlign.CENTER,
      margins: { top: 60, bottom: 60, left: 100, right: 100 },
      children: [
        new Paragraph({
          alignment: align,
          children: toRuns(parseInline(text), {
            bold: header,
            size: 18,
            color: header ? "FFFFFF" : undefined,
          }),
        }),
      ],
    });

  return new Table({
    width: { size: CONTENT_WIDTH, type: WidthType.DXA },
    columnWidths: widths,
    borders: {
      top: { style: BorderStyle.SINGLE, size: 2, color: "B7C6E5" },
      bottom: { style: BorderStyle.SINGLE, size: 2, color: "B7C6E5" },
      left: { style: BorderStyle.SINGLE, size: 2, color: "B7C6E5" },
      right: { style: BorderStyle.SINGLE, size: 2, color: "B7C6E5" },
      insideHorizontal: { style: BorderStyle.SINGLE, size: 2, color: "D7DFF2" },
      insideVertical: { style: BorderStyle.SINGLE, size: 2, color: "D7DFF2" },
    },
    rows: [
      new TableRow({
        tableHeader: true,
        children: headerCells.map((h, idx) =>
          cell(h, { header: true, align: AlignmentType.CENTER, width: widths[idx] })
        ),
      }),
      ...bodyRows.map(
        (r) =>
          new TableRow({
            children: r.map((c, idx) =>
              cell(c, { align: aligns[idx] || AlignmentType.LEFT, width: widths[idx] })
            ),
          })
      ),
    ],
  });
}

// ---------------------------------------------------------------------------
// Corps du document
// ---------------------------------------------------------------------------

/** Le corps démarre au premier titre de section : l'en-tête sert la couverture. */
let startIdx = 0;
for (let i = 0; i < lines.length; i += 1) {
  if (/^#{1,2} (Partie|Sommaire|\d+\.)/.test(lines[i])) {
    startIdx = i;
    break;
  }
}

/** Champ « **Nom :** valeur » de l'en-tête. */
function headerField(name) {
  const m = new RegExp(`\\*\\*${name}\\s*:\\*\\*\\s*([^\\n]+)`).exec(raw.slice(0, 2500));
  return m ? `${name} : ${m[1].replace(/\*\*/g, "").trim()}` : null;
}

/** Avertissement de tête : les lignes « > » situées avant la première section. */
function coverQuotes() {
  return lines
    .slice(0, startIdx)
    .filter((l) => l.startsWith(">"))
    .map((l) => l.replace(/^>\s?/, "").trim())
    .filter(Boolean);
}

const body = [];
let blockquote = [];

function flushBlockquote() {
  for (const l of blockquote) {
    body.push(
      new Paragraph({
        indent: { left: 360 },
        border: { left: { style: BorderStyle.SINGLE, size: 18, color: BRAND2, space: 8 } },
        shading: { type: ShadingType.CLEAR, fill: LIGHT },
        spacing: { after: 80 },
        children: toRuns(parseInline(l), { italics: true, size: 20, color: "3B4A66" }),
      })
    );
  }
  blockquote = [];
}

let i = startIdx;
while (i < lines.length) {
  const line = lines[i];

  if (line.trim() === "" || line.trim() === "---") {
    flushBlockquote();
    i += 1;
    continue;
  }

  if (line.trim() === "```") {
    flushBlockquote();
    i += 1;
    const code = [];
    while (i < lines.length && lines[i].trim() !== "```") {
      code.push(lines[i]);
      i += 1;
    }
    i += 1;
    code.forEach((cl, idx) =>
      body.push(
        new Paragraph({
          shading: { type: ShadingType.CLEAR, fill: "F2F2F2" },
          spacing: { after: idx === code.length - 1 ? 160 : 0 },
          children: [new TextRun({ text: cl, font: "Consolas", size: 18 })],
        })
      )
    );
    continue;
  }

  if (line.startsWith(">")) {
    blockquote.push(line.replace(/^>\s?/, ""));
    i += 1;
    continue;
  }
  flushBlockquote();

  const heading = /^(#{1,4})\s+(.*)$/.exec(line);
  if (heading) {
    const level = heading[1].length;
    const text = heading[2].trim();
    const levels = {
      1: HeadingLevel.HEADING_1,
      2: HeadingLevel.HEADING_2,
      3: HeadingLevel.HEADING_3,
      4: HeadingLevel.HEADING_4,
    };
    body.push(
      new Paragraph({
        heading: levels[level],
        pageBreakBefore: level === 1 && /^Partie /.test(text),
        spacing: { before: level === 1 ? 400 : 240, after: 160 },
        children: toRuns(parseInline(text)),
      })
    );
    i += 1;
    continue;
  }

  if (line.trim().startsWith("|")) {
    const tbl = [];
    while (i < lines.length && lines[i].trim().startsWith("|")) {
      tbl.push(lines[i]);
      i += 1;
    }
    if (tbl.length >= 2) {
      body.push(
        buildTable(splitRow(tbl[0]), splitRow(tbl[1]).map(parseAlign), tbl.slice(2).map(splitRow))
      );
      body.push(new Paragraph({ text: "", spacing: { after: 160 } }));
    }
    continue;
  }

  // Une puce ou un élément numéroté garde sa marque : la conserver en texte
  // évite une numérotation Word qui repartirait à 1 à chaque liste.
  const listItem = /^(\s*)([-*]|\d+\.)\s+(.*)$/.exec(line);
  if (listItem) {
    const marker = /^\d/.test(listItem[2]) ? `${listItem[2]} ` : "• ";
    body.push(
      new Paragraph({
        indent: { left: 360, hanging: 200 },
        spacing: { after: 80 },
        children: toRuns(parseInline(marker + listItem[3])),
      })
    );
    i += 1;
    continue;
  }

  body.push(
    new Paragraph({ spacing: { after: 140 }, children: toRuns(parseInline(line.trim())) })
  );
  i += 1;
}
flushBlockquote();

// ---------------------------------------------------------------------------
// Page de garde
// ---------------------------------------------------------------------------

const centered = (text, opts) =>
  new Paragraph({
    alignment: AlignmentType.CENTER,
    spacing: opts.spacing,
    children: [new TextRun({ text, font: FONT, ...opts.run })],
  });

const cover = [
  new Paragraph({ spacing: { before: 1200 }, text: "" }),
  centered(COVER.title, {
    spacing: { after: 120 },
    run: { bold: true, size: 52, color: BRAND },
  }),
];
if (COVER.subtitle) {
  cover.push(
    centered(COVER.subtitle, {
      spacing: { before: 120, after: 100 },
      run: { bold: true, size: 30, color: BRAND2 },
    })
  );
}
if (COVER.tagline) {
  cover.push(
    centered(COVER.tagline, {
      spacing: { after: 400 },
      run: { italics: true, size: 24, color: "44546A" },
    })
  );
}
for (const field of ["Public", "Destinataires", "Durée de présentation conseillée", "Version", "Date"]) {
  const value = headerField(field);
  if (value) {
    cover.push(centered(value, { spacing: { after: 40 }, run: { size: 20, color: "44546A" } }));
  }
}
const classification = headerField("Classification");
cover.push(
  centered(classification ?? "Classification : usage interne", {
    spacing: { after: 420 },
    run: { size: 20, italics: true, color: "44546A" },
  })
);
for (const quote of coverQuotes()) {
  cover.push(
    new Paragraph({
      indent: { left: 360, right: 360 },
      border: { left: { style: BorderStyle.SINGLE, size: 18, color: BRAND2, space: 8 } },
      shading: { type: ShadingType.CLEAR, fill: LIGHT },
      spacing: { after: 160 },
      children: toRuns(parseInline(quote), { italics: true, size: 20, color: "3B4A66" }),
    })
  );
}
cover.push(new Paragraph({ children: [new PageBreak()] }));

// ---------------------------------------------------------------------------
// Document
// ---------------------------------------------------------------------------

const headingStyle = (id, name, size, color, extra = {}) => ({
  id,
  name,
  basedOn: "Normal",
  next: "Normal",
  quickFormat: true,
  run: { font: FONT, size, bold: true, color },
  paragraph: { spacing: { before: 280, after: 140 }, keepNext: true, ...extra },
});

const doc = new Document({
  creator: "Direction des Risques",
  title: COVER.title,
  description: COVER.subtitle,
  styles: {
    default: { document: { run: { font: FONT, size: 21 } } },
    paragraphStyles: [
      headingStyle("Heading1", "Heading 1", 30, BRAND, { spacing: { before: 360, after: 180 } }),
      headingStyle("Heading2", "Heading 2", 25, BRAND2, {
        border: { bottom: { style: BorderStyle.SINGLE, size: 6, color: "C3CFE8", space: 4 } },
      }),
      headingStyle("Heading3", "Heading 3", 22, "31456A"),
      {
        ...headingStyle("Heading4", "Heading 4", 21, "44546A"),
        run: { font: FONT, size: 21, bold: true, italics: true, color: "44546A" },
      },
    ],
  },
  sections: [
    {
      properties: {
        page: {
          size: { width: 11906, height: 16838 }, // A4
          margin: { top: 1000, bottom: 1000, left: 1150, right: 1150 },
        },
      },
      headers: {
        default: new Header({
          children: [
            new Paragraph({
              tabStops: [{ type: TabStopType.RIGHT, position: TabStopPosition.MAX }],
              border: { bottom: { style: BorderStyle.SINGLE, size: 4, color: "C3CFE8", space: 4 } },
              children: [
                new TextRun({ text: COVER.header, size: 15, color: "6B7A99", font: FONT }),
                new TextRun({
                  text: `\t${COVER.version.split(" · ")[0]}`,
                  size: 15,
                  color: "6B7A99",
                  font: FONT,
                }),
              ],
            }),
          ],
        }),
      },
      footers: {
        default: new Footer({
          children: [
            new Paragraph({
              alignment: AlignmentType.CENTER,
              children: [
                new TextRun({ text: "Usage interne — ", size: 15, color: "6B7A99", font: FONT }),
                new TextRun({ children: [PageNumber.CURRENT], size: 15, color: "6B7A99", font: FONT }),
                new TextRun({ text: " / ", size: 15, color: "6B7A99", font: FONT }),
                new TextRun({ children: [PageNumber.TOTAL_PAGES], size: 15, color: "6B7A99", font: FONT }),
              ],
            }),
          ],
        }),
      },
      children: [...cover, ...body],
    },
  ],
});

const buffer = await Packer.toBuffer(doc);
fs.mkdirSync(path.dirname(OUT), { recursive: true });
fs.writeFileSync(OUT, buffer);
console.log(`${path.relative(process.cwd(), OUT)} — ${(buffer.length / 1024).toFixed(0)} Ko`);
